### ISSUE TITLE
Fix response to AskForBlocksInfo::Operations

### ISSUE THREAD

**Issue by gterzian** [massalabs/massa#2884]
* no need to perform this `filter`
* first make sure that when the `AskForBlocksInfo::Operations` message is deserialized, we fail if there are too many ops being asked (constant config parameter)
* before we do `send_block_info` on response to `AskForBlocksInfo::Operations` we should first gather all the asked operations from storage, and if any is absent, send NotFound
* if all are found, send all the WrappedOperations throught send_block_info (and not just their IDs)
* make sure the order of operations in the block are preserved

_Originally posted by @damip in https://github.com/massalabs/massa/pull/2862#discussion_r946954656_

**Comment by gterzian** [comment]
@damip I have a few questions:

> send all the WrappedOperations throught send_block_info (and not just their IDs)

Currently, the idea is that protocol sends a message to network saying "send those operations", using the IDs, and network will retrieve the serialized form of the op from storage when sending the message. 

Note that network doesn't need the wrapped op, just the serialized form. 

> make sure the order of operations in the block are preserved

The current idea is that the operations and the block are separated, so the "order" in which operations are received from the network seems irrelevant. The only thing that matter, I think, is:

1. We store a list of operation for the block(in order). 
2. We have all those operations present in storage. 

So passing around a `Vec<OperationId>` in the (network) message, instead of a set, does not seem to contributed to the above goals. 

> we should first gather all the asked operations from storage, and if any is absent, send NotFound

If a node replies to another node with info on the block, it has the block, so it should also have all the operations as well.  

> no need to perform this filter

Actually I think we should strengthen it and make sure we ops are part of the block, to avoid having a node ask for nonsense and reply with potentially non-existing ops(which would fail when network would try to retrieve them from storage). 


Re all of the above, we can change the current workflow and we should clarify exactly what we want to do here.

**Comment by gterzian** [comment]
It should work something like this:

1. Node A has a block(integrated it). Therefore it also has the ordered list of ops.
2. Node A propagates header.
3. Node B receives header.
4. Node B adds block to wishlist, asks Node A for info.
5. Node A sends `BlockInfoReply::Info(list)`, where `list` is the ordered list of ops. 
6. Node B asks for a set of missing ops.
7. Node A sends the set of missing ops. 

So basically, `BlockInfoReply::Info(list)` should contain the ordered list, while the rest of the workflow to get the actual missing ops does not need an ordered list.

**Comment by damip** [comment]
@gterzian 

> @damip I have a few questions:
> 
> > send all the WrappedOperations throught send_block_info (and not just their IDs)
> 
> Currently, the idea is that protocol sends a message to network saying "send those operations", using the IDs, and network will retrieve the serialized form of the op from storage when sending the message.

I think that the best option would be to send an instance of Storage that holds references to those operations. That way we avoid cases where operations disappear from storage while the message is in flight.

Do you think that we should also do it the other way around ? (eg. storing incoming objects to storage within Network, and then sending a Storage to Protocol) 

> 
> Note that network doesn't need the wrapped op, just the serialized form.

If we just send an instance of Storage, Network will be able to clone the serialized version from storage directly and take only what it needs.

> 
> > make sure the order of operations in the block are preserved
> 
> The current idea is that the operations and the block are separated, so the "order" in which operations are received from the network seems irrelevant. The only thing that matter, I think, is:
> 
>     1. We store a list of operation for the block(in order).
> 
>     2. We have all those operations present in storage.
> 
> 
> So I'm wondering how passing around a `Vec<OperationId>` in the (network) messages, instead of a set, does contribute to the above goals.

True.

One thing that might be useful is to ensure that while we are waiting for missing operations in a block, we hold on to the ones we already have using a Storage specific to that block (and that block Storage instance would eventually become the Storage we send to Consensus when the block is complete, or that we drop if we cancel the block retrieval).

> 
> > we should first gather all the asked operations from storage, and if any is absent, send NotFound
> 
> If a node replies to another node with info on the block, it has the block, so it should also have all the operations as well.
> 

Here I am talking about the case where OUR node receives `AskForBlocksInfo::Operations`. It either has ALL the asked ops in storage and sends them, or it should send NotFound.

> > no need to perform this filter
> 
> Actually I think we should strengthen it and make sure we ops are part of the block, to avoid having a node ask for nonsense and reply with potentially non-existing ops(which would fail when network would try to retrieve them from storage).

Ok 

> 
> Re all of the above, we can change the current workflow and we should clarify exactly what we want to do here.

It should be rather unambiguous now

Note that those comment were not processed:
* https://github.com/massalabs/massa/pull/2862#discussion_r946955873
* https://github.com/massalabs/massa/pull/2862#discussion_r946958282

**Comment by damip** [comment]
> It should work something like this:
> 
>     1. Node A has a wrapped block in storage. Therefore it also has the ordered list of ops.
> 
>     2. Node A propagates header.
> 
>     3. Node B receives header.
> 
>     4. Node B adds block to wishlist, asks Node A for info.
> 
>     5. Node A sends `BlockInfoReply::Info(list)`, where `list` is the ordered list of ops.
> 
>     6. Node B asks for a set of missing ops.
> 
>     7. Node A sends the set of missing ops.
> 
> 
> So basically, `BlockInfoReply::Info(list)` should contain the ordered list, while the rest of the workflow to get the actual missing ops does not need an ordered list.

True

> 
> In the light of #2883, we probably also need to use storage inside protocol in such a way as to guarantee that the block/ops remain in storage from 1 to 7.

Yes, that's the point of https://github.com/massalabs/massa/pull/2888

**Comment by damip** [comment]
@gterzian any progress on this issue ?

**Comment by gterzian** [comment]
@gterzian any progress on this issue ?

I think the most important things were done in https://github.com/massalabs/massa/pull/2891, what remains is the stuff related of storage discussed above. 

I am currently focusing on fixing the protocol tests, so maybe someone else can do it. @adrien-zinger ? I believe all info you need is in the comments of @damip above.

**Comment by damip** [comment]
ok let's close it
